exporter/otlpexporter/internal/arrow/stream.go

```diff
@@ -0,0 +1,246 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package arrow
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	arrowpb "github.com/f5/otel-arrow-adapter/api/collector/arrow/v1"
+	arrowRecord "github.com/f5/otel-arrow-adapter/pkg/otel/arrow_record"
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+
+	"go.opentelemetry.io/collector/consumer/consumererror"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+)
+
+// Stream is 1:1 with gRPC stream.
+type Stream struct {
+	// client uses the exporter's grpc.ClientConn.
+	client arrowpb.ArrowStreamService_ArrowStreamClient
+
+	// toWrite is passes a batch from the sender to the stream writer, which
+	// includes a dedicated channel for the response.
+	toWrite chan writeItem
+
+	// producer is exclusive to the holder of the stream.
+	producer *arrowRecord.Producer
+
+	// cancel cancels the stream context.
+	cancel context.CancelFunc
+
+	// lock protects waiters.
+	lock sync.Mutex
+
+	// waiters is the response channel for each active batch.
+	waiters map[string]chan error
+}
+
+// writeItem is passed from the sender (a pipeline consumer) to the
+// stream writer, which is not bound by the sender's context.
+type writeItem struct {
+	// records is a ptrace.Traces, plog.Logs, or pmetric.Metrics
```

Comment:
question: is there some interface that all the pdata adhere to? i would love to avoid interface{} where possible.

Reply:
The interface can become a generic `plog.Logs | pmetric.Metrics | ptrace.Traces`, an interface otel-arrow-adapter package is exporting (as of a recent PR). I will push this update into the next PR.